File: lib/clusterbuster/pod_files/sync.pl
#!/usr/bin/perl
use Socket;
use POSIX;
use strict;
use Time::Piece;
use Time::HiRes qw(gettimeofday usleep);
use JSON;
use Scalar::Util qw(looks_like_number);

my ($verbose, $sync_file, $error_file, $controller_timestamp_file);
my ($offset_from_controller) = 0;
my ($predelay) = 0;
my ($postdelay) = 0;
use Getopt::Long;
Getopt::Long::Configure("bundling", "no_ignore_case", "pass_through");
GetOptions("v!"  => \$verbose,
	   "t:s" => \$controller_timestamp_file,
	   "d:i" => \$predelay,
	   "D:i" => \$postdelay,
	   "f:s" => \$sync_file,
	   "e:s" => \$error_file);

my ($listen_port, $expected_clients, $initial_expected_clients, $sync_count) = @ARGV;
if ($sync_count < 1 || $expected_clients < 1) {
    timestamp("Sync requested with no clients or no syncs");
    POSIX::exit(0);
}

if ($initial_expected_clients <= 0) {
    $initial_expected_clients = $expected_clients;
}

sub xtime() {
    my (@now) = gettimeofday();
    return $now[0] + ($now[1] / 1000000.0);
}

sub clean_numbers($) {
    # Perl to_json encodes innfinity as inf and NaN as nan.
    # This results in invalid JSON.  It's our responsibility to sanitize
    # this up front.
    my ($ref) = @_;
    if (ref $ref eq 'HASH') {
	my (%answer);
	map { $answer{$_} = clean_numbers($$ref{$_})} keys %$ref;
	return \%answer;
    } elsif (ref $ref eq 'ARRAY') {
	my (@answer) = map {clean_numbers($_)} @$ref;
	return \@answer;
    } elsif (ref $ref eq '' && looks_like_number($ref) && $ref != 0 &&
	     (! defined ($ref <=> 0) || ((1 / $ref) == 0))) {
	return undef;
    } else {
	return $ref
    }
}

sub touch($) {
    my ($file) = @_;
    open FILE, ">", $file || die "Can't open $file: $!\n";
    close FILE || die "Can't close $file: $!\n";
}
sub ytime() {
    return xtime() + $offset_from_controller;
}

sub timestamp($) {
    my ($str) = @_;
    my (@now) = POSIX::modf(ytime());
    printf STDERR  "sync %s.%06d %s\n", gmtime(int($now[1]))->strftime("%Y-%m-%dT%T"), $now[0] * 1000000, $str;
}

sub read_token($) {
    my ($client) = @_;
    my ($tbuf) = '';
    if (sysread($client, $tbuf, 10) != 10) {
	timestamp("Unable to read token");
	return undef;
    }
    if (!($tbuf =~ /0x[[:xdigit:]]{8}/)) {
	timestamp("Bad token $tbuf!");
	return undef;
    }
    my ($bytes_to_read) = hex($tbuf);
    $tbuf = '';
    my ($offset) = 0;
    while ($bytes_to_read > 0) {
	my ($bytes) = sysread($client, $tbuf, $bytes_to_read, $offset);
	if ($bytes == 0) {
	    timestamp("Short read: got zero bytes with $bytes_to_read left at $offset");
	    return undef;
	} elsif ($bytes < 0) {
	    timestamp("Bad read with $bytes_to_read left at $offset: $!");
	    return undef;
	} else {
	    $bytes_to_read -= $bytes;
	    $offset += $bytes;
	}
    }
    return $tbuf
}

sub get_port($) {
    my ($listen_port) = @_;
    my $sockaddr = "S n a4 x8";
    my $sock;
    socket($sock, AF_INET, SOCK_STREAM, getprotobyname('tcp')) || die "socket: $!";
    $SIG{TERM} = sub { close SOCK; POSIX::_exit(0); };
    setsockopt($sock,SOL_SOCKET, SO_REUSEADDR, pack("l",1)) || die "setsockopt reuseaddr: $!\n";
    setsockopt($sock,SOL_SOCKET, SO_KEEPALIVE, pack("l",1)) || die "setsockopt keepalive: $!\n";
    bind($sock, pack($sockaddr, AF_INET, $listen_port, "\0\0\0\0")) || die "bind: $!\n";
    my $mysockaddr = getsockname($sock);
    my ($junk, $port, $addr) = unpack($sockaddr, $mysockaddr);
    die "can't get port $port: $!\n" if ($port ne $listen_port);
    return $sock;
}

sub get_controller_timing($) {
    # Normalize time to the run host.  We collect two timestamps on the run host
    # bracketing one taken here, giving us an approximate delta between the two
    # hosts.  We assume that the second timestamp on the host is taken closer to
    # the sync host timestamp on the grounds that setting up the oc exec
    # is slower than tearing it down, but in either event, we know what the worst
    # case error is.
    timestamp("Waiting to receive controller time data");
    my ($timestamp_file) = @_;
    if ($timestamp_file) {
	while (! -f $timestamp_file) {
	    usleep(100000);
	}
	open TIMESTAMP, "<", $timestamp_file || die "Can't open timestamp file $timestamp_file: $!\n";
	my ($controller_json_data) = "";
	while (<TIMESTAMP>) {
	    $controller_json_data .= $_;
	}
	close TIMESTAMP;
	timestamp("Timestamp data: $controller_json_data");
	my ($controller_timestamp_data) = from_json($controller_json_data);
	$$controller_timestamp_data{'offset_from_controller'} = $$controller_timestamp_data{'second_controller_ts'} - $$controller_timestamp_data{'sync_ts'};
	unlink($timestamp_file);
	return $controller_timestamp_data;
    }
}

sub handle_rslt($$$) {
    my ($tmp_sync_file_base, $expected_clients, $tbuf) = @_;
    if (defined $tmp_sync_file_base) {
	my ($tmp_sync_file) = sprintf("%s-%d", $tmp_sync_file_base, $expected_clients);
	chomp $tbuf;
	open TMP, ">", "$tmp_sync_file" || die("Can't open sync file $tmp_sync_file: $!\n");
	print TMP "$tbuf\n";
	close TMP || die "Can't close sync file: $!\n";
    }
}

sub reply_timestamp($$\@) {
    my ($start_time, $base_start_time, $ts_clients) = @_;
    my ($start) = ytime();
    timestamp("Returning client sync start time, sync start time, sync sent time");
    foreach my $client (@$ts_clients) {
	my ($time) = ytime();
	my ($client_fd, $client_ts) = @$client;
	my ($tbuf) = "$client_ts $start_time $start $time $base_start_time";
	syswrite($client_fd, $tbuf, length $tbuf);
	close($client_fd);
    }
    my ($end) = ytime();
    my ($et) = $end - $start;
    timestamp("Sending sync time took $et seconds");
}

sub sync_one($$$$$$$) {
    my ($sock, $tmp_sync_file_base, $tmp_error_file, $start_time, $base_start_time, $expected_clients, $first_pass) = @_;
    timestamp("Listening on port $listen_port");
    listen($sock, $expected_clients) || die "listen: $!";
    printf STDERR "Expect $expected_clients client%s\n", $expected_clients == 1 ? '' : 's';
    my (@ts_clients);
    # Ensure that the client file descriptors do not get gc'ed,
    # closing it prematurely.  This is used when we don't
    # need to send a meaningful reply.
    # Tested with
    # clusterbuster -P synctest --synctest-count=10 --synctest-cluster-count=3 --precleanup
    #     --deployments=10 --cleanup=0 --pin-node=whatever
    my (@clients_protect_against_gc);
    while ($expected_clients > 0) {
	my ($client);
	accept($client, $sock) || next;
	my $peer = getpeername($client);
	my ($port, $addr) = sockaddr_in($peer);
	# Reverse hostname lookup adds significant overhead
	# when using sync to establish the timebase.
	#my $peerhost = gethostbyaddr($addr, AF_INET);
	my $peeraddr = inet_ntoa($addr);
	my ($tbuf) = read_token($client);
	if (! defined $tbuf) {
	    timestamp("Read token from $peeraddr failed: $!");
	}
	push @clients_protect_against_gc, $client;
	my ($command) = lc substr($tbuf, 0, 4);
	my ($payload_bytes) = length $tbuf - 4;
	timestamp("Accepted connection from $peeraddr on $port, command $command, payload $payload_bytes");
	$tbuf =~ s/^....\s+//;
	if ($command eq 'time')  {
	    my ($ignore, $ts, $ignore) = split(/ +/, $tbuf);
	    if (! $first_pass) {
		timestamp("Unexpected request for time sync from $tbuf!");
		open TMP, ">", "$tmp_error_file" || die ("Can't open error file $tmp_error_file: $!\n");
		print TMP "Unexpected request for time sync from $tbuf!";
		close TMP || die "Can't close error file: $!\n";
		link($tmp_error_file, $error_file) || die "Can't link $tmp_error_file to $error_file: $!\n";
		timestamp("Waiting for error file $error_file to be removed");
		while (-f $error_file) {
		    sleep(1);
		}
		POSIX::_exit(1)
	    }
	    push @ts_clients, [$client, "$ts " . ytime()];
	} elsif ($command eq 'rslt') {
	    handle_rslt($tmp_sync_file_base, $expected_clients, $tbuf);
	} elsif ($command eq 'fail') {
	    timestamp("Detected failure from $peeraddr");
	    if (defined $tmp_error_file) {
		open TMP, ">", "$tmp_error_file" || die("Can't open error file $tmp_error_file: $!\n");
		print TMP "$tbuf\n";
		close TMP || die "Can't close error file: $!\n";
		link($tmp_error_file, $error_file) || die "Can't link $tmp_error_file to $error_file: $!\n";
		timestamp("Waiting for error file $error_file to be removed");
		while (-f $error_file) {
		    sleep(1);
		}
	    } else {
		timestamp("Message: $tbuf");
	    }
	    POSIX::_exit(1);
	} elsif ($command eq 'sync') {
	    # No special handling needed for a normal sync
	} else {
	    timestamp("Unknown command from $peeraddr: '$command'");
	}
	$expected_clients--;
    }
    # Only reply to clients who provided a timestamp
    if (@ts_clients) {
	reply_timestamp($start_time, $base_start_time, @ts_clients);
    }
    timestamp("Sync complete");
    POSIX::_exit(0);
}

my ($start_time) = xtime();
my ($base_start_time) = $start_time;

$SIG{TERM} = sub { POSIX::_exit(0); };

my ($original_sync_count) = $sync_count;

timestamp("Clusterbuster sync starting");
my ($sock) = get_port($listen_port);

my ($tmp_sync_file_base) = (defined($sync_file) && $sync_file ne '') ? "${sync_file}-tmp" : undef;
my ($tmp_error_file) = (defined($error_file) && $error_file ne '') ? "${error_file}-tmp" : undef;

my (@tmp_sync_files) = map { "${tmp_sync_file_base}-$_" } (1..$expected_clients);

my ($controller_timestamp_data) = get_controller_timing($controller_timestamp_file);
timestamp("About to adjust timestamp");
timestamp(sprintf("Max timebase error %f" , $$controller_timestamp_data{'offset_from_controller'}));
$start_time += $$controller_timestamp_data{'offset_from_controller'};
timestamp("Adjusted timebase by $$controller_timestamp_data{'offset_from_controller'} seconds $base_start_time => $start_time");
timestamp("Will sync $sync_count times");
if ($sync_count == 0) {
    timestamp("No synchronization requested; sleeping $postdelay seconds");
    sleep($postdelay);
} else {
    my ($first_pass) = 1;
    while ($sync_count < 0 || $sync_count-- > 0) {
	if (-e $tmp_error_file) {
	    timestamp("Job failed, exiting");
	    exit(1);
	}
	my ($tmp_sync_file) = undef;
	# Ensure that all of the accepted connections get closed by exiting
	# a child process.  This way we don't have to keep track of all of the
	# clients and close them manually.
	my $child = fork();
	if ($child == 0) {
	    sync_one($sock, $tmp_sync_file_base, $tmp_error_file, $start_time, $base_start_time, $first_pass ? $initial_expected_clients : $expected_clients, $first_pass);
	} elsif ($child < 1) {
	    timestamp("Fork failed: $!");
	    POSIX::_exit(1);
	} else {
	    wait();
	}
	if ($first_pass) {
	    touch("/tmp/clusterbuster-started");
	    if ($predelay > 0) {
		timestamp("Waiting $predelay seconds before start");
		sleep($predelay);
	    }
	    $first_pass = 0;
	}
    }
    touch("/tmp/clusterbuster-finished");
    if ($postdelay > 0) {
	timestamp("Waiting $postdelay seconds before end");
	sleep($postdelay);
    }
}

if (-e $tmp_error_file) {
    timestamp("Job failed, exiting!");
    POSIX::_exit(1);
}
my (%result);
$result{'controller_timing'} = $controller_timestamp_data;
my (@data);

if (@tmp_sync_files) {
    foreach my $file (@tmp_sync_files) {
	if (-f $file && open FILE, "<", $file) {
	    my ($datum) = "";
	    while (<FILE>) {
		$datum .= $_;
	    }
	    close FILE;
	    push @data, from_json($datum);
	} else {
	    push @data, {};
	}
    }
} elsif ($original_sync_count == 0) {
    @data = map { {} } (1..$expected_clients);
    timestamp("here");
    timestamp(join("|", @data));
} else {
    timestamp("orig sync count $original_sync_count");
}
$result{'worker_results'} = \@data;

open (TMP, ">", $tmp_sync_file_base) || die "Can't open sync file $tmp_sync_file_base: $!\n";
print TMP to_json(clean_numbers(\%result));
close TMP || die "Can't close temporary sync file: $!\n";
rename($tmp_sync_file_base, $sync_file) || die "Can't rename $tmp_sync_file_base to $sync_file: $!\n";
timestamp("Waiting for sync file $sync_file to be removed");
while (-f $sync_file) {
    sleep(1);
}
timestamp("Sync file $sync_file removed, exiting");

POSIX::_exit(0);
EOF

File: lib/clusterbuster/pod_files/sync.py
#!/usr/bin/env python3

# Copyright 2022 Robert Krawitz/Red Hat
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import time
import re
import json
import os
import sys
from cb_util import cb_util

offset_from_controller = 0
timebase = cb_util(offset_from_controller)
nameserver_pid = None
sync_nonce = None


def kill_nameserver(timebase: cb_util):
    global nameserver_pid
    if nameserver_pid:
        timebase._timestamp("Killing nameserver")
        os.kill(nameserver_pid, 9)


def fatal(string: str):
    timebase._timestamp(string)
    kill_nameserver
    sys.exit(1)


def ytime():
    global offset_from_controller
    return time.time() + offset_from_controller


def touch(file: str):
    with open(file, "w") as f:
        f.write('')


class nameserver:
    """
    Simplistic nameserver for ClusterBuster workloads
    """
    def __init__(self, timebase: cb_util, port: int, backlog: int = 5):
        self.timebase = timebase
        self.addrs = dict()
        self.requests = dict()
        self.clients = dict()
        try:
            self.sock = timebase._listen(port=port, backlog=backlog)
        except Exception as err:
            self.fatal(f"listen failed: {err}")

    def timestamp(self, string):
        self.timebase._timestamp("Nameserver: " + str(string))

    def fatal(self, string: str):
        self.timestamp(string)
        sys.exit(1)

    def satisfy_requests(self):
        requestors_satisfied = list()
        for addr, req in self.requests.items():
            requests_satisfied = list()
            for want in req['want'].keys():
                if want in self.addrs:
                    self.timestamp(f"    Request from {addr} for {want} => {self.addrs[want]}")
                    self.requests[addr]['have'][want] = self.addrs[want]
                    requests_satisfied.append(want)
            if requests_satisfied:
                for want in requests_satisfied:
                    del self.requests[addr]['want'][want]
                if not self.requests[addr]['want']:
                    jdata = json.dumps(self.requests[addr]['have'])
                    self.timestamp(f"        All requests from {addr} are satisfied: {jdata}, sending")
                    self.clients[addr].send(jdata.encode('ascii'))
                    self.clients[addr].close()
                    del self.clients[addr]
                    requestors_satisfied.append(addr)
        if requestors_satisfied:
            for requestor in requestors_satisfied:
                del self.requests[requestor]

    def get_command(self):
        client, address = self.sock.accept()
        try:
            tbuf = read_token(client)
            nonce, tbuf = tbuf.split(' ', 1)
            if nonce != sync_nonce:
                self.timestamp(f"Received request with incorrect nonce {nonce} from {address}: {tbuf}")
                return None, None, None
            command = tbuf[0:4].lower()
            payload = tbuf[4:].lstrip()
            self.timestamp(f"Accepted connection from {address}, command {command}, payload {payload}")
            json_payload = json.loads(payload)
            if command != 'nsrq':
                raise ValueError(f"Unexpected command {command}")
        except Exception as exc:
            self.timestamp(f"Could not read command from {client} at {address}: {exc}")
            client.close()
            return
        return client, address, json_payload

    def process_command(self):
        try:
            client, address, json_payload = self.get_command()
            if client is None and address is None and json_payload is None:
                return
        except Exception as exc:
            self.timestamp(f"Unable to read command: {exc}")
            return
        for command, args in json_payload.items():
            if command == 'have':
                if not isinstance(args, dict):
                    self.timestamp(f"have payload should be dict, is {args}")
                    continue
                for name, ipaddr in args.items():
                    self.timestamp(f"    {address} offers {name} at {ipaddr}")
                    self.addrs[name] = ipaddr
            elif command == 'rqst':
                if not isinstance(args, list):
                    self.timestamp(f"rqst payload should be list, is {args}")
                    continue
                if address not in self.requests:
                    self.requests[address] = {
                        'have': {},
                        'want': {}
                        }
                for name in args:
                    if address not in self.clients:
                        self.clients[address] = client
                    self.requests[address]['want'][name] = None
            else:
                self.timestamp(f"Unknown command from {address}: '{command}'")

    def run(self):
        while True:
            self.process_command()
            self.satisfy_requests()


def read_token(stream):
    prefix = stream.recv(10).decode('ascii').lower()
    if len(prefix) == 0:
        return None
    if len(prefix) != 10:
        raise ValueError("Unable to read token: short read")
    m = re.match(r'0x[0-9a-z]{8}', prefix)
    if m:
        bytes_to_read = int(prefix, base=16)
    else:
        raise ValueError(f"Bad token: {prefix}")
    answer = ''
    offset = 0
    while bytes_to_read > 0:
        try:
            chunk = stream.recv(bytes_to_read).decode('ascii')
        except Exception as exc:
            raise ValueError(f"Bad read with {bytes_to_read} left at offset {offset}: {exc}")
        nbytes = len(chunk)
        if nbytes == 0:
            raise ValueError(f"Short read: got zero bytes with {bytes_to_read} left at {offset}")
        else:
            answer += chunk
            bytes_to_read -= nbytes
            offset += nbytes
    return answer


def get_controller_timing(timestamp_file: str):
    """
    Normalize time to the run host.  We collect two timestamps on the run host
    bracketing one taken here, giving us an approximate delta between the two
    hosts.  We assume that the second timestamp on the host is taken closer to
    the sync host timestamp on the grounds that setting up the oc exec
    is slower than tearing it down, but in either event, we know what the worst
    case error is.
    """
    while not timebase._isfile(timestamp_file):
        time.sleep(0.1)
    with open(timestamp_file, "r", encoding='ascii') as tsfile:
        try:
            controller_json_data = tsfile.read()
        except Exception as error:
            fatal(f"Cannot read data from {tsfile}: {error}")
    timebase._timestamp(f"Timestamp data: {controller_json_data}")
    try:
        tsdata = json.loads(controller_json_data)
        tsdata['offset_from_controller'] = tsdata['second_controller_ts'] - tsdata['sync_ts']
    except Exception as error:
        fatal(f"Cannot convert data to JSON: {error}")
    try:
        os.unlink(timestamp_file)
    except Exception:
        pass
    return tsdata


def handle_result(tmp_sync_file_base: str, expected_clients: int, tbuf: str):
    tmp_sync_file = f'{tmp_sync_file_base}-{expected_clients}'
    try:
        with open(tmp_sync_file, 'w') as tmp:
            tmp.write(tbuf.rstrip())
    except Exception as error:
        fatal(f"Can't write to sync file {tmp_sync_file}: {error}")


def reply_timestamp(ts_clients: list):
    start = ytime()
    timebase._timestamp("Returning client sync start time, sync start time, sync sent time")
    for client in ts_clients:
        client_fd, client_ts = client
        client_ts['reply_start'] = start
        client_ts['reply_time'] = ytime()
        client_fd.send(json.dumps(client_ts).encode('ascii'))
    end = ytime()
    et = end - start
    timebase._timestamp(f"Sending sync time took {et} seconds")


def fail_hard(payload: str):
    if tmp_error_file:
        try:
            with open(tmp_error_file, "w") as tmp:
                tmp.write(payload)
        except Exception as err:
            fatal(f"Can't write to {tmp_error_file}: {err}")
        try:
            os.link(tmp_error_file, error_file)
        except Exception as err:
            fatal(f"Can't link {tmp_error_file} to {error_file}: {err}")
        timebase._timestamp(f"Waiting for error file {error_file} to be removed")
        while timebase._isfile(error_file):
            time.sleep(1)
    else:
        timebase._timestamp("Message: {payload}")
    os._exit(1)


def sync_one(sock, tmp_sync_file_base: str, tmp_error_file: str, start_time: float,
             base_start_time: float, expected_clients: int, first_pass: bool):
    timebase._timestamp(f"Listening on port {listen_port}, expect {expected_clients} client{'' if expected_clients == 1 else 's'}")
    try:
        timebase._listen(sock=sock, backlog=expected_clients)
    except Exception as err:
        fatal(f"listen failed: {err}")
    ts_clients = []
    net_clients = {}
    # Ensure that the client file descriptors do not get gc'ed,
    # closing it prematurely.  This is used when we don't
    # need to send a meaningful reply.  Without this, we do get some
    # failures.
    # Tested with
    # clusterbuster -P synctest --synctest-count=1000 --synctest-cluster-count=3
    #     --precleanup --deployments=10 --cleanup=0
    protected_clients = []
    expected_command = None
    while expected_clients > 0:
        # Reverse hostname lookup adds significant overhead
        # when using sync to establish the timebase.
        client, address = sock.accept()
        tbuf = read_token(client)
        if not tbuf:
            timebase._timestamp(f"Read token from {address} failed")
            continue
        try:
            nonce, tbuf = tbuf.split(' ', 1)
        except Exception as exc:
            timebase._timestamp(f"Could not parse token {tbuf}: {exc}")
            continue
        # Don't acknowledge replies with incorrect nonce
        if nonce != sync_nonce:
            timebase._timestamp(f"Received request with incorrect nonce {nonce} from {address}: {tbuf}")
            continue
        protected_clients.append(client)
        command = tbuf[0:4].lower()
        if expected_command:
            if command != expected_command:
                fatal(f"Unexpected command {command} from {address}, expected {expected_command}")
        else:
            expected_command = command
        payload = tbuf[4:].lstrip()
        timebase._timestamp(f"Accepted connection from {address}, command {command}, payload {len(payload)}")
        if command == 'time' or command == 'tnet':
            timebase._timestamp(f"Time request {payload}")
            if first_pass:
                if command == 'tnet':
                    try:
                        jdata = json.loads(payload)
                        ts = jdata['timestamp']
                        if 'have' in jdata and isinstance(jdata['have'], dict):
                            for name, addr in jdata['have'].items():
                                if name.startswith('eth0@'):
                                    timebase._timestamp(f"Using {address[0]} for {name}")
                                    net_clients[name] = address[0]
                                else:
                                    net_clients[name] = addr
                    except Exception as exc:
                        timebase._timestamp(f"Failed to parse JSON data: {exc}")
                        continue
                else:
                    ignore, ts, ignore = payload.split()
                ts_clients.append([client,
                                   {
                                    "client_ts": ts,
                                    "request_time": ytime(),
                                    "start_time": start_time,
                                    "base_start_time": base_start_time
                                    }])
            else:
                fail_hard(f"Unexpected request for time sync from {payload}")
        elif command == 'rslt':
            handle_result(tmp_sync_file_base, expected_clients, payload)
        elif command == 'fail':
            timebase._timestamp(f"Detected failure from {address}")
            fail_hard(payload)
        elif command == 'sync':
            if step_interval > 0:
                timebase._timestamp(f"Waiting for {step_interval} seconds...")
                time.sleep(step_interval)
                timebase._timestamp("Done waiting")
        else:
            timebase._timestamp(f"Unknown command from {address}: '{command}'")
        expected_clients -= 1
    if ts_clients:
        reply_timestamp(ts_clients)
    if net_clients:
        msg = json.dumps({'have': net_clients})
        command = f'nsrq {msg}'
        timebase._send_message('127.0.0.1', ns_port, f"{sync_nonce} {command}")
    timebase._timestamp("Sync complete")
    if command == 'rslt':
        return 2
    else:
        return 0


print(sys.argv, file=sys.stderr)
try:
    sync_nonce = sys.argv[1]
    sync_file = sys.argv[2]
    error_file = sys.argv[3]
    controller_timestamp_file = sys.argv[4]
    predelay = float(sys.argv[5])
    postdelay = float(sys.argv[6])
    step_interval = float(sys.argv[7])
    listen_port = int(sys.argv[8])
    ns_port = int(sys.argv[9])
    expected_clients = int(sys.argv[10])
    initial_expected_clients = int(sys.argv[11])
    if initial_expected_clients < 0:
        initial_expected_clients = expected_clients
except Exception as exc:
    timebase._timestamp(f"Can't initialize arguments: {exc}")

start_time = time.time()
base_start_time = start_time
timebase._timestamp("Clusterbuster sync starting")
sock = timebase._get_port(addr=None, port=listen_port)
if sync_file:
    tmp_sync_file_base = f'{sync_file}-tmp'
else:
    tmp_sync_file_base = None
if error_file:
    tmp_error_file = f'{error_file}-tmp'
else:
    tmp_error_file = None
tmp_sync_files = [f"{tmp_sync_file_base}-{i}" for i in range(1, expected_clients + 1)]

controller_timestamp_data = get_controller_timing(controller_timestamp_file)
offset_from_controller = controller_timestamp_data['offset_from_controller']

timebase._timestamp("About to adjust timestamp")
timebase._timestamp(f"Max timebase error {offset_from_controller}")
timebase._set_offset(-offset_from_controller)
start_time += offset_from_controller
timebase._timestamp(f"Adjusted timebase by {offset_from_controller} seconds {base_start_time} => {start_time}")
try:
    child = os.fork()
except Exception as exc:
    fatal(f"Fork failed: {exc}")
if child == 0:
    timebase._timestamp("About to launch nameserver")
    nameserver = nameserver(timebase, ns_port, expected_clients).run()
    sys.exit()
else:
    nameserver_pid = child
timebase._timestamp("Starting sync")
first_pass = True
while True:
    if timebase._isfile(tmp_error_file):
        fatal("Job failed, exiting")
    tmp_sync_file = None
    # Ensure that all of the accepted connections get closed by exiting
    # a child process.  This way we don't have to keep track of all of the
    # clients and close them manually.
    try:
        child = os.fork()
    except Exception as exc:
        fatal(f"Fork failed: {exc}")
    status = -1
    if child == 0:
        if first_pass:
            clients = initial_expected_clients
        else:
            clients = expected_clients
        sys.exit(sync_one(sock, tmp_sync_file_base, tmp_error_file, start_time, base_start_time, clients, first_pass))
    else:
        try:
            pid, status = os.wait()
        except Exception as err:
            fatal(f"Wait failed: {err}")
    if first_pass:
        touch("/tmp/clusterbuster-started")
        if predelay > 0:
            timebase._timestamp(f"Waiting {predelay} seconds before start")
            time.sleep(predelay)
        first_pass = False
    if (status >> 8) == 2:
        timebase._timestamp("Final sync complete, finishing up")
        touch("/tmp/clusterbuster-finished")
        break
    elif status & 255:
        fatal(f"Sync killed by signal {status & 255}")
    elif status != 0:
        fatal("Job failed, exiting")

if postdelay > 0:
    timebase._timestamp(f"Waiting {postdelay} seconds before end")
    time.sleep(postdelay)

if timebase._isfile(tmp_error_file):
    fatal("Job failed, exiting")

result = {
    'controller_timing': controller_timestamp_data
    }

data = []

if tmp_sync_files:
    for f in tmp_sync_files:
        try:
            with open(f, "r") as fp:
                content = fp.read()
                datum = json.loads(content)
                data.append(datum)
        except Exception as exc:
            timebase._timestamp(f"Could not load JSON from {f}: {exc}")
            data.append(dict())
else:
    data = [dict() for i in expected_clients]
result['worker_results'] = data

try:
    with open(tmp_sync_file_base, 'w') as tmp:
        tmp.write(json.dumps(timebase._clean_numbers(result), sort_keys=True, indent=1))
except Exception as exc:
    fatal(f"Can't write to sync file {tmp_sync_file_base}: {exc}")
try:
    os.rename(tmp_sync_file_base, sync_file)
except Exception as exc:
    fatal(f"Can't rename {tmp_sync_file_base} to {sync_file}: {exc}")
timebase._timestamp(f"Waiting for sync file {sync_file} to be removed")
while timebase._isfile(sync_file):
    time.sleep(1)
timebase._timestamp(f"Sync file {sync_file} removed, exiting")
kill_nameserver(timebase)
sys.exit(0)
